File: flutter_app/lib/services/assistant/offline_tools.dart
import 'dart:math' as math;

const _dayNamesDe = ['Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag', 'Sonntag'];
const _monthNamesDe = ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni',
    'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember'];

String _fmtNum(double n) {
  if ((n - n.roundToDouble()).abs() < 1e-9) return n.round().toString();
  var s = n.toStringAsFixed(6);
  while (s.endsWith('0')) {
    s = s.substring(0, s.length - 1);
  }
  if (s.endsWith('.')) s = s.substring(0, s.length - 1);
  return s.replaceAll('.', ',');
}

final _calcPrefixes = RegExp(
  r'^\s*(?:(?:bitte\s+)?(?:ich\s+)?(?:berechne|rechne|löse|loese|was\s+ist|was\s+ergibt|wie\s*viel\s+(?:ist|sind|ergibt)|wieviel\s+(?:ist|sind|ergibt))\s+)+',
  caseSensitive: false,
);
final _calcStripDecor = RegExp(r'[^\w\d+\-*/().,%^\s√]', unicode: true);

String? _evalArith(String expr) {
  expr = expr.replaceAll(' ', '').replaceAll(',', '.').replaceAll('^', '**');
  if (!RegExp(r'^[\d+\-*/().%]+$').hasMatch(expr.replaceAll('**', '*'))) return null;
  if (expr.length < 2) return null;
  try {
    return _evalSafe(expr).toString();
  } catch (_) {
    return null;
  }
}

// Tiny shunting-yard evaluator — supports + - * / ** ( ) and a single leading minus.
double _evalSafe(String expr) {
  final tokens = <String>[];
  var i = 0;
  while (i < expr.length) {
    final c = expr[i];
    if (RegExp(r'[\d.]').hasMatch(c)) {
      final start = i;
      while (i < expr.length && RegExp(r'[\d.]').hasMatch(expr[i])) {
        i++;
      }
      tokens.add(expr.substring(start, i));
      continue;
    }
    if (c == '*' && i + 1 < expr.length && expr[i + 1] == '*') {
      tokens.add('**');
      i += 2;
      continue;
    }
    if ('+-*/()'.contains(c)) {
      // Handle leading unary minus / after open paren
      if (c == '-' && (tokens.isEmpty || tokens.last == '(' || RegExp(r'^[-+*/]$|^\*\*$').hasMatch(tokens.last))) {
        tokens.add('0');
      }
      tokens.add(c);
      i++;
      continue;
    }
    throw const FormatException('bad char');
  }
  // Convert to RPN (shunting-yard)
  const prec = {'+': 1, '-': 1, '*': 2, '/': 2, '**': 3};
  final output = <String>[];
  final opStack = <String>[];
  for (final t in tokens) {
    if (RegExp(r'^\d').hasMatch(t)) {
      output.add(t);
    } else if (t == '(') {
      opStack.add(t);
    } else if (t == ')') {
      while (opStack.isNotEmpty && opStack.last != '(') {
        output.add(opStack.removeLast());
      }
      if (opStack.isEmpty) throw const FormatException('mismatched paren');
      opStack.removeLast();
    } else {
      while (opStack.isNotEmpty && opStack.last != '(' &&
             ((prec[opStack.last] ?? 0) > (prec[t] ?? 0) ||
              ((prec[opStack.last] ?? 0) == (prec[t] ?? 0) && t != '**'))) {
        output.add(opStack.removeLast());
      }
      opStack.add(t);
    }
  }
  while (opStack.isNotEmpty) {
    final op = opStack.removeLast();
    if (op == '(') throw const FormatException('mismatched paren');
    output.add(op);
  }
  // Eval RPN
  final stack = <double>[];
  for (final t in output) {
    if (RegExp(r'^\d').hasMatch(t)) {
      stack.add(double.parse(t));
    } else {
      if (stack.length < 2) throw const FormatException('underflow');
      final b = stack.removeLast();
      final a = stack.removeLast();
      switch (t) {
        case '+': stack.add(a + b); break;
        case '-': stack.add(a - b); break;
        case '*': stack.add(a * b); break;
        case '/':
          if (b == 0) throw const FormatException('div0');
          stack.add(a / b);
          break;
        case '**': stack.add(math.pow(a, b).toDouble()); break;
        default: throw const FormatException('unknown op');
      }
    }
  }
  if (stack.length != 1) throw const FormatException('incomplete');
  return stack.single;
}

String? tryCalc(String message) {
  var raw = message.trim().replaceAll(RegExp(r'[?.!]+$'), '').trim();
  raw = raw.replaceAll(_calcStripDecor, ' ').trim();
  final m = _calcPrefixes.allMatches(raw).isEmpty ? raw : raw.replaceFirst(_calcPrefixes, '').trim();
  if (m.isEmpty) return null;

  // Percentage
  final pct = RegExp(r'^(\d+(?:[.,]\d+)?)\s*(?:%|prozent)\s+von\s+(\d+(?:[.,]\d+)?)', caseSensitive: false).firstMatch(m);
  if (pct != null) {
    final a = double.parse(pct.group(1)!.replaceAll(',', '.'));
    final b = double.parse(pct.group(2)!.replaceAll(',', '.'));
    return '${_fmtNum(a)} % von ${_fmtNum(b)} = **${_fmtNum(a * b / 100)}**';
  }

  // sqrt
  var root = RegExp(r'^(?:die\s+)?wurzel\s+(?:aus\s+)?(\d+(?:[.,]\d+)?)', caseSensitive: false).firstMatch(m);
  root ??= RegExp(r'^sqrt\s*\(?\s*(\d+(?:[.,]\d+)?)\s*\)?', caseSensitive: false).firstMatch(m);
  if (root != null) {
    final n = double.parse(root.group(1)!.replaceAll(',', '.'));
    if (n < 0) return null;
    return '√${_fmtNum(n)} = **${_fmtNum(math.sqrt(n))}**';
  }

  // Word operators
  final word = RegExp(r'^(\-?\d+(?:[.,]\d+)?)\s+(mal|plus|minus|durch|geteilt durch|hoch)\s+(\-?\d+(?:[.,]\d+)?)', caseSensitive: false).firstMatch(m);
  if (word != null) {
    final a = double.parse(word.group(1)!.replaceAll(',', '.'));
    final b = double.parse(word.group(3)!.replaceAll(',', '.'));
    final op = word.group(2)!.toLowerCase();
    double res;
    switch (op) {
      case 'mal': res = a * b; break;
      case 'plus': res = a + b; break;
      case 'minus': res = a - b; break;
      case 'durch':
      case 'geteilt durch':
        if (b == 0) return 'Teilen durch 0 ist nicht definiert.';
        res = a / b;
        break;
      case 'hoch': res = math.pow(a, b).toDouble(); break;
      default: return null;
    }
    return '${_fmtNum(a)} $op ${_fmtNum(b)} = **${_fmtNum(res)}**';
  }

  // Bare arithmetic expression
  if (RegExp(r'^[\d\s+\-*/().,%^]+$').hasMatch(m)) {
    final hasOp = RegExp(r'[+\-*/^%]').hasMatch(m);
    if (hasOp && m.length >= 2) {
      final result = _evalArith(m);
      if (result != null) {
        final val = double.tryParse(result);
        if (val != null) {
          final pretty = m.replaceAll('**', '^').replaceAll('*', '·').replaceAll('/', ' ÷ ');
          return '$pretty = **${_fmtNum(val)}**';
        }
      }
    }
  }
  return null;
}

// ---- Date math ----

DateTime? _parseDateDe(String s, DateTime now) {
  var m = RegExp(r'(\d{1,2})\.\s*(\d{1,2})\.\s*(\d{2,4})').firstMatch(s);
  if (m != null) {
    var d = int.parse(m.group(1)!);
    var mo = int.parse(m.group(2)!);
    var y = int.parse(m.group(3)!);
    if (y < 100) y += 2000;
    try {
      final dt = DateTime(y, mo, d);
      if (dt.month != mo || dt.day != d) return null;
      return dt;
    } catch (_) {
      return null;
    }
  }
  final words = _monthNamesDe.map((n) => n.toLowerCase()).join('|');
  m = RegExp(r'(\d{1,2})\.\s*(' + words + r')(?:\s+(\d{2,4}))?', caseSensitive: false).firstMatch(s);
  if (m != null) {
    final d = int.parse(m.group(1)!);
    final mo = _monthNamesDe.indexWhere((n) => n.toLowerCase() == m!.group(2)!.toLowerCase()) + 1;
    var y = m.group(3) != null ? int.parse(m.group(3)!) : now.year;
    if (y < 100) y += 2000;
    try {
      var dt = DateTime(y, mo, d);
      if (dt.month != mo || dt.day != d) return null;
      if (m.group(3) == null && dt.isBefore(DateTime(now.year, now.month, now.day))) {
        dt = DateTime(y + 1, mo, d);
      }
      return dt;
    } catch (_) {
      return null;
    }
  }
  return null;
}

String _dayName(DateTime dt) => _dayNamesDe[dt.weekday - 1];

String _fmtDateDe(DateTime d) =>
    '${d.day.toString().padLeft(2, '0')}.${d.month.toString().padLeft(2, '0')}.${d.year}';

String? tryDateMath(String message, {DateTime? now}) {
  final n = now ?? DateTime.now();
  final today = DateTime(n.year, n.month, n.day);
  final m = message.trim().toLowerCase().replaceAll(RegExp(r'[?.!]+$'), '').trim();

  if (RegExp(r'^(?:welcher|was für ein)\s+(wochen)?tag\s+ist\s+heute').hasMatch(m)) {
    return 'Heute ist **${_dayName(today)}**, der ${_fmtDateDe(today)}.';
  }
  if (RegExp(r'^(?:welcher|was für ein)\s+(wochen)?tag\s+ist\s+morgen').hasMatch(m)) {
    final t = today.add(const Duration(days: 1));
    return 'Morgen ist **${_dayName(t)}**, der ${_fmtDateDe(t)}.';
  }
  final wt = RegExp(r'welcher\s+(?:wochen)?tag(?:\s+der\s+woche)?\s+ist\s+(?:der|am)?\s*(.+)').firstMatch(m);
  if (wt != null) {
    final dt = _parseDateDe(wt.group(1)!, n);
    if (dt != null) return 'Der ${_fmtDateDe(dt)} ist ein **${_dayName(dt)}**.';
  }
  final bis = RegExp(r'wie\s*viele\s+tage\s+(?:bis|noch\s+bis|bis\s+zum?)\s+(.+)').firstMatch(m);
  if (bis != null) {
    final dt = _parseDateDe(bis.group(1)!, n);
    if (dt != null) {
      final diff = dt.difference(today).inDays;
      if (diff == 0) return 'Das ist heute.';
      if (diff < 0) return 'Das Datum (${_fmtDateDe(dt)}) liegt **${-diff} Tage** zurück.';
      return 'Bis zum ${_fmtDateDe(dt)} sind es noch **$diff Tage**.';
    }
  }
  final inX = RegExp(r'(?:was für ein|welcher)\s+(?:wochen)?tag\s+ist\s+in\s+(\d+)\s+(tag|tage|woche|wochen|monat|monate)').firstMatch(m);
  if (inX != null) {
    final count = int.parse(inX.group(1)!);
    final unit = inX.group(2)!;
    DateTime dt;
    if (unit.contains('tag')) {
      dt = today.add(Duration(days: count));
    } else if (unit.contains('woche')) {
      dt = today.add(Duration(days: count * 7));
    } else {
      final totalMonths = today.month + count;
      final y = today.year + ((totalMonths - 1) ~/ 12);
      final mo = ((totalMonths - 1) % 12) + 1;
      dt = DateTime(y, mo, today.day);
    }
    return 'In $count $unit ist **${_dayName(dt)}**, der ${_fmtDateDe(dt)}.';
  }
  if (RegExp(r'wie\s+sp[äa]t(\s+ist(\s+es)?)?|(?:was\s+ist\s+die|welche)\s+uhrzeit|^uhrzeit$|uhrzeit\s+bitte|sag\s+mir\s+die\s+uhrzeit').hasMatch(m)) {
    return 'Es ist **${n.hour.toString().padLeft(2, '0')}:${n.minute.toString().padLeft(2, '0')}** Uhr.';
  }
  return null;
}

// ---- Unit conversion ----

const _unitAliases = {
  'km': 'km',
  'kilometer': 'km', 'kilometers': 'km',
  'meile': 'mi', 'meilen': 'mi', 'mi': 'mi', 'mile': 'mi', 'miles': 'mi',
  'm': 'm', 'meter': 'm', 'meters': 'm',
  'ft': 'ft', 'fuß': 'ft', 'fuss': 'ft', 'foot': 'ft', 'feet': 'ft',
  'kg': 'kg', 'kilogramm': 'kg', 'kilogram': 'kg', 'kilograms': 'kg',
  'pfund': 'lb', 'lb': 'lb', 'lbs': 'lb', 'pound': 'lb', 'pounds': 'lb',
  'l': 'l', 'liter': 'l', 'litre': 'l', 'litres': 'l', 'liters': 'l',
  'gallon': 'gal', 'gallonen': 'gal', 'gallons': 'gal', 'gal': 'gal',
};

const _unitConversions = {
  'km>mi': [1 / 1.609344, 'Meilen'],
  'mi>km': [1.609344, 'km'],
  'm>ft':  [3.28084, 'Fuß'],
  'ft>m':  [0.3048, 'm'],
  'm>km':  [0.001, 'km'],
  'km>m':  [1000.0, 'm'],
  'kg>lb': [2.20462, 'Pfund'],
  'lb>kg': [0.453592, 'kg'],
  'l>gal': [0.264172, 'Gallonen'],
  'gal>l': [3.78541, 'l'],
};

String? tryUnitConvert(String message) {
  final m = message.trim().toLowerCase().replaceAll(RegExp(r'[?.!]+$'), '').trim();

  final temp = RegExp(
    r'^(\-?\d+(?:[.,]\d+)?)\s*(?:grad\s+)?(celsius|c|fahrenheit|f|kelvin|k)\s+(?:in|nach|zu|to)\s+(?:grad\s+)?(celsius|c|fahrenheit|f|kelvin|k)',
    caseSensitive: false,
  ).firstMatch(m);
  if (temp != null) {
    final val = double.parse(temp.group(1)!.replaceAll(',', '.'));
    final src = temp.group(2)!.toLowerCase()[0];
    final dst = temp.group(3)!.toLowerCase()[0];
    double res;
    if (src == 'c' && dst == 'f') {
      res = val * 9 / 5 + 32;
    } else if (src == 'f' && dst == 'c') {
      res = (val - 32) * 5 / 9;
    } else if (src == 'c' && dst == 'k') {
      res = val + 273.15;
    } else if (src == 'k' && dst == 'c') {
      res = val - 273.15;
    } else if (src == 'f' && dst == 'k') {
      res = (val - 32) * 5 / 9 + 273.15;
    } else if (src == 'k' && dst == 'f') {
      res = (val - 273.15) * 9 / 5 + 32;
    } else {
      res = val;
    }
    const labels = {'c': '°C', 'f': '°F', 'k': 'K'};
    return '${_fmtNum(val)} ${labels[src]} = **${_fmtNum(res)} ${labels[dst]}**';
  }

  final conv = RegExp(
    r'^(\-?\d+(?:[.,]\d+)?)\s*([a-zäöüß]+)\s+(?:in|nach|zu|to)\s+([a-zäöüß]+)',
    caseSensitive: false,
  ).firstMatch(m);
  if (conv != null) {
    final val = double.parse(conv.group(1)!.replaceAll(',', '.'));
    final srcRaw = conv.group(2)!.toLowerCase();
    final dstRaw = conv.group(3)!.toLowerCase();
    final src = _unitAliases[srcRaw];
    final dst = _unitAliases[dstRaw];
    if (src != null && dst != null) {
      final entry = _unitConversions['$src>$dst'];
      if (entry != null) {
        final factor = entry[0] as double;
        final label = entry[1] as String;
        return '${_fmtNum(val)} $srcRaw = **${_fmtNum(val * factor)} $label**';
      }
    }
  }
  return null;
}

String? tryOfflineTool(String message) {
  if (message.isEmpty || message.length > 300) return null;
  for (final fn in <String? Function(String)>[tryDateMath, tryUnitConvert, tryCalc]) {
    try {
      final out = fn(message);
      if (out != null) return out;
    } catch (_) {
      continue;
    }
  }
  return null;
}

File: app/offline_tools.py
from __future__ import annotations

import ast
import math
import re
from datetime import datetime, timedelta
from typing import Optional

DAY_NAMES_DE = ['Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag', 'Sonntag']
MONTH_NAMES_DE = ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni',
                  'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember']

_NUMBER_WORDS_DE = {
    'null': 0, 'eins': 1, 'eine': 1, 'einen': 1, 'zwei': 2, 'drei': 3, 'vier': 4,
    'fünf': 5, 'sechs': 6, 'sieben': 7, 'acht': 8, 'neun': 9, 'zehn': 10,
    'elf': 11, 'zwölf': 12, 'zwanzig': 20, 'dreißig': 30, 'hundert': 100, 'tausend': 1000,
}


def _parse_num(s: str) -> Optional[float]:
    s = s.strip().lower().replace(',', '.').replace(' ', '')
    if s in _NUMBER_WORDS_DE:
        return float(_NUMBER_WORDS_DE[s])
    try:
        return float(s)
    except ValueError:
        return None


def _fmt_num(n: float) -> str:
    if abs(n - round(n)) < 1e-9:
        return str(int(round(n)))
    s = f'{n:.6f}'.rstrip('0').rstrip('.')
    return s.replace('.', ',')


# ---- Math ----

_SAFE_MATH = re.compile(r'^[\d\s+\-*/().,%^√πesqrtinlogcosabc\u00d7\u00f7]+$')
_MATH_TRIGGERS = re.compile(
    r'^(?:(?:berechne|was\s+ist|wie\s+viel\s+ist|wieviel\s+ist|rechne|löse)\s+)?'
    r'(.+?)'
    r'(?:\s*=\s*\?)?$',
    re.IGNORECASE,
)


_CALC_PREFIXES = re.compile(
    r'^\s*(?:(?:bitte\s+)?(?:ich\s+)?(?:berechne|rechne|löse|loese|was\s+ist|was\s+ergibt|wie\s*viel\s+(?:ist|sind|ergibt)|wieviel\s+(?:ist|sind|ergibt))\s+)+',
    re.IGNORECASE,
)
_CALC_STRIP_TRAILING = re.compile(r'[^\w\d+\-*/().,%^\s√]', re.UNICODE)

_MAX_POW_EXPONENT = 1000
_MAX_POW_BASE = 10 ** 18


def _safe_eval_arith(expr: str):
    """Evaluate a bare arithmetic expression without eval().

    Supports + - * / % ** and parentheses over numeric literals. Exponentiation
    is bounded so a short input like '9**9**9' cannot trigger a big-integer DoS.
    """
    def _node(n):
        if isinstance(n, ast.Expression):
            return _node(n.body)
        if isinstance(n, ast.Constant) and isinstance(n.value, (int, float)) and not isinstance(n.value, bool):
            return n.value
        if isinstance(n, ast.UnaryOp) and isinstance(n.op, (ast.UAdd, ast.USub)):
            v = _node(n.operand)
            return +v if isinstance(n.op, ast.UAdd) else -v
        if isinstance(n, ast.BinOp):
            left = _node(n.left)
            right = _node(n.right)
            op = n.op
            if isinstance(op, ast.Add):
                return left + right
            if isinstance(op, ast.Sub):
                return left - right
            if isinstance(op, ast.Mult):
                return left * right
            if isinstance(op, ast.Div):
                return left / right
            if isinstance(op, ast.Mod):
                return left % right
            if isinstance(op, ast.Pow):
                if abs(right) > _MAX_POW_EXPONENT or abs(left) > _MAX_POW_BASE:
                    raise ValueError('exponent out of bounds')
                return left ** right
        raise ValueError('unsupported expression')

    return _node(ast.parse(expr, mode='eval'))


def try_calc(message: str) -> Optional[str]:
    raw = message.strip().rstrip('?.!').strip()
    # Drop decorative noise (emoji, arrows, …) that can't belong to arithmetic.
    raw = _CALC_STRIP_TRAILING.sub(' ', raw).strip()
    m = _CALC_PREFIXES.sub('', raw).strip()
    if not m:
        return None

    # Percentage: "15% von 200", "wie viel sind 15 prozent von 200"
    pct = re.match(r'(\d+(?:[.,]\d+)?)\s*(?:%|prozent)\s+von\s+(\d+(?:[.,]\d+)?)', m, re.IGNORECASE)
    if pct:
        a = float(pct.group(1).replace(',', '.'))
        b = float(pct.group(2).replace(',', '.'))
        return f'{_fmt_num(a)} % von {_fmt_num(b)} = **{_fmt_num(a * b / 100)}**'

    # Square root: "wurzel aus 16", "sqrt(144)", "sqrt 25"
    root = re.match(r'(?:die\s+)?wurzel\s+(?:aus\s+)?(\d+(?:[.,]\d+)?)', m, re.IGNORECASE)
    if not root:
        root = re.match(r'sqrt\s*\(?\s*(\d+(?:[.,]\d+)?)\s*\)?', m, re.IGNORECASE)
    if root:
        n = float(root.group(1).replace(',', '.'))
        if n < 0:
            return None
        return f'√{_fmt_num(n)} = **{_fmt_num(math.sqrt(n))}**'

    # "2 mal 8", "7 plus 3", "20 minus 5", "50 durch 4", "3 hoch 4"
    word_op = re.match(r'(\-?\d+(?:[.,]\d+)?)\s+(mal|plus|minus|durch|geteilt durch|hoch)\s+(\-?\d+(?:[.,]\d+)?)', m, re.IGNORECASE)
    if word_op:
        a = float(word_op.group(1).replace(',', '.'))
        b = float(word_op.group(3).replace(',', '.'))
        op = word_op.group(2).lower()
        try:
            if op == 'mal':
                res = a * b
            elif op == 'plus':
                res = a + b
            elif op == 'minus':
                res = a - b
            elif op in ('durch', 'geteilt durch'):
                if b == 0:
                    return 'Teilen durch 0 ist nicht definiert.'
                res = a / b
            elif op == 'hoch':
                res = a ** b
            else:
                return None
            return f'{_fmt_num(a)} {op} {_fmt_num(b)} = **{_fmt_num(res)}**'
        except Exception:
            return None

    # Bare arithmetic expression: "2+2", "3*(4+5)", "2^10"
    expr_match = re.match(r'^([\d\s+\-*/().,%^]+)$', m)
    if expr_match:
        expr = expr_match.group(1).replace(' ', '').replace(',', '.').replace('^', '**')
        if not re.fullmatch(r'[\d+\-*/().%]+', expr.replace('**', '*')):
            return None
        if len(expr) < 2 or '+' not in expr and '-' not in expr and '*' not in expr and '/' not in expr and '%' not in expr and '**' not in expr:
            return None
        try:
            result = _safe_eval_arith(expr)
            if isinstance(result, (int, float)):
                pretty_expr = expr.replace('**', '^').replace('*', '·').replace('/', ' ÷ ').replace('+', ' + ').replace('-', ' − ')
                return f'{pretty_expr.strip()} = **{_fmt_num(float(result))}**'
        except Exception:
            return None
    return None


# ---- Date math ----

_DATE_DE = re.compile(r'(\d{1,2})\.\s*(\d{1,2})\.\s*(\d{2,4})')
_DATE_WORD = re.compile(r'(\d{1,2})\.\s*(' + '|'.join(MONTH_NAMES_DE) + r')(?:\s+(\d{2,4}))?', re.IGNORECASE)


def _parse_date_de(s: str, now: Optional[datetime] = None) -> Optional[datetime]:
    now = now or datetime.now()
    m = _DATE_DE.search(s)
    if m:
        d, mo, y = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if y < 100:
            y += 2000
        try:
            return datetime(y, mo, d)
        except ValueError:
            return None
    m = _DATE_WORD.search(s)
    if m:
        d = int(m.group(1))
        mo = MONTH_NAMES_DE.index(m.group(2).capitalize()) + 1
        y = int(m.group(3)) if m.group(3) else now.year
        if y < 100:
            y += 2000
        try:
            dt = datetime(y, mo, d)
            if not m.group(3) and dt < now:
                dt = dt.replace(year=y + 1)
            return dt
        except ValueError:
            return None
    return None


def try_date_math(message: str) -> Optional[str]:
    m = message.strip().lower().rstrip('?.!').strip()
    now = datetime.now()

    if re.match(r'^(?:welcher|was für ein)\s+(wochen)?tag\s+ist\s+heute\b', m):
        return f'Heute ist **{DAY_NAMES_DE[now.weekday()]}**, der {now.strftime("%d.%m.%Y")}.'
    if re.match(r'^(?:welcher|was für ein)\s+(wochen)?tag\s+ist\s+morgen\b', m):
        t = now + timedelta(days=1)
        return f'Morgen ist **{DAY_NAMES_DE[t.weekday()]}**, der {t.strftime("%d.%m.%Y")}.'

    # "welcher wochentag / tag der woche ist der 24.12.2026?"
    wt = re.search(r'welcher\s+(?:wochen)?tag(?:\s+der\s+woche)?\s+ist\s+(?:der|am)?\s*(.+)', m)
    if wt:
        dt = _parse_date_de(wt.group(1), now)
        if dt:
            return f'Der {dt.strftime("%d.%m.%Y")} ist ein **{DAY_NAMES_DE[dt.weekday()]}**.'

    # "wie viele tage bis zum 01.06.2026"
    days_to = re.search(r'wie\s*viele\s+tage\s+(?:bis|noch\s+bis|bis\s+zum?)\s+(.+)', m)
    if days_to:
        dt = _parse_date_de(days_to.group(1), now)
        if dt:
            diff = (dt.date() - now.date()).days
            if diff == 0:
                return 'Das ist heute.'
            if diff < 0:
                return f'Das Datum ({dt.strftime("%d.%m.%Y")}) liegt **{-diff} Tage** zurück.'
            return f'Bis zum {dt.strftime("%d.%m.%Y")} sind es noch **{diff} Tage**.'

    # "in 3 wochen", "in 10 tagen"
    in_x = re.search(r'(?:was für ein|welcher)\s+(?:wochen)?tag\s+ist\s+in\s+(\d+)\s+(tag|tage|woche|wochen|monat|monate)', m)
    if in_x:
        n = int(in_x.group(1))
        unit = in_x.group(2)
        if 'tag' in unit:
            dt = now + timedelta(days=n)
        elif 'woche' in unit:
            dt = now + timedelta(weeks=n)
        else:
            mo = now.month + n
            y = now.year + (mo - 1) // 12
            mo = ((mo - 1) % 12) + 1
            try:
                dt = now.replace(year=y, month=mo)
            except ValueError:
                dt = now.replace(year=y, month=mo, day=1)
        return f'In {n} {unit} ist **{DAY_NAMES_DE[dt.weekday()]}**, der {dt.strftime("%d.%m.%Y")}.'

    # "wie spät", "Uhrzeit?", "welche uhrzeit haben wir", "sag mir die uhrzeit"
    if re.search(r'wie\s+sp[äa]t(\s+ist(\s+es)?)?|(?:was\s+ist\s+die|welche)\s+uhrzeit|^uhrzeit$|uhrzeit\s+bitte|sag\s+mir\s+die\s+uhrzeit', m):
        return f'Es ist **{now.strftime("%H:%M")}** Uhr.'

    return None


# ---- Unit conversion ----

_UNIT_ALIASES = {
    'km': 'km',
    'kilometer': 'km', 'kilometers': 'km',
    'meile': 'mi', 'meilen': 'mi', 'mi': 'mi', 'mile': 'mi', 'miles': 'mi',
    'm': 'm', 'meter': 'm', 'meters': 'm',
    'ft': 'ft', 'fuß': 'ft', 'fuss': 'ft', 'foot': 'ft', 'feet': 'ft',
    'kg': 'kg', 'kilogramm': 'kg', 'kilogram': 'kg', 'kilograms': 'kg',
    'pfund': 'lb', 'lb': 'lb', 'lbs': 'lb', 'pound': 'lb', 'pounds': 'lb',
    'l': 'l', 'liter': 'l', 'litre': 'l', 'litres': 'l', 'liters': 'l',
    'gallon': 'gal', 'gallonen': 'gal', 'gallons': 'gal', 'gal': 'gal',
}

_UNIT_CONVERSIONS = {
    ('km', 'mi'): (1 / 1.609344, 'Meilen'),
    ('mi', 'km'): (1.609344, 'km'),
    ('m',  'ft'): (3.28084, 'Fuß'),
    ('ft', 'm'):  (0.3048, 'm'),
    ('m',  'km'): (0.001, 'km'),
    ('km', 'm'):  (1000.0, 'm'),
    ('kg', 'lb'): (2.20462, 'Pfund'),
    ('lb', 'kg'): (0.453592, 'kg'),
    ('l',  'gal'): (0.264172, 'Gallonen'),
    ('gal', 'l'):  (3.78541, 'l'),
}


def try_unit_convert(message: str) -> Optional[str]:
    m = message.strip().lower().rstrip('?.!').strip()

    temp = re.match(r'(\-?\d+(?:[.,]\d+)?)\s*(?:grad\s+)?(celsius|c|fahrenheit|f|kelvin|k)\s+(?:in|nach|zu|to)\s+(?:grad\s+)?(celsius|c|fahrenheit|f|kelvin|k)', m, re.IGNORECASE)
    if temp:
        val = float(temp.group(1).replace(',', '.'))
        src = temp.group(2).lower()[0]
        dst = temp.group(3).lower()[0]
        if src == 'c' and dst == 'f':
            res = val * 9 / 5 + 32
        elif src == 'f' and dst == 'c':
            res = (val - 32) * 5 / 9
        elif src == 'c' and dst == 'k':
            res = val + 273.15
        elif src == 'k' and dst == 'c':
            res = val - 273.15
        elif src == 'f' and dst == 'k':
            res = (val - 32) * 5 / 9 + 273.15
        elif src == 'k' and dst == 'f':
            res = (val - 273.15) * 9 / 5 + 32
        else:
            res = val
        unit_long = {'c': '°C', 'f': '°F', 'k': 'K'}
        return f'{_fmt_num(val)} {unit_long[src]} = **{_fmt_num(res)} {unit_long[dst]}**'

    conv = re.match(r'(\-?\d+(?:[.,]\d+)?)\s*([a-zäöüß]+)\s+(?:in|nach|zu|to)\s+([a-zäöüß]+)', m, re.IGNORECASE)
    if conv:
        val = float(conv.group(1).replace(',', '.'))
        src_raw = conv.group(2).lower()
        dst_raw = conv.group(3).lower()
        src = _UNIT_ALIASES.get(src_raw)
        dst = _UNIT_ALIASES.get(dst_raw)
        if src and dst:
            key = (src, dst)
            if key in _UNIT_CONVERSIONS:
                factor, label = _UNIT_CONVERSIONS[key]
                return f'{_fmt_num(val)} {src_raw} = **{_fmt_num(val * factor)} {label}**'

    return None


# ---- Top-level dispatcher ----

def try_offline_tool(message: str) -> Optional[str]:
    if not message or len(message) > 200:
        return None
    for fn in (try_date_math, try_unit_convert, try_calc):
        try:
            result = fn(message)
            if result:
                return result
        except Exception:
            continue
    return None
